### PR TITLE
vol_p val→test gap: domain-adversarial backbone training (v2)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The val→test volume pressure gap (+7–8pp, persistent across 15+ falsified interventions) is a **covariate shift problem**: train/val car configurations have systematically different aerodynamic volume pressure distributions than test configurations. Every model-side intervention tried so far has failed because none of them address the root cause — the backbone learns train-distribution-specific features that don't generalize to the test distribution.

**Domain adversarial training (DANN)** forces the backbone to produce distribution-invariant features via a gradient-reversal discriminator that cannot distinguish train-vs-test domain. A Gradient-Reversal Layer (GRL) is inserted between the backbone and a domain discriminator. During backprop, the discriminator tries to identify the domain; the GRL negates its gradient so the backbone is pushed toward domain-invariant representations. If successful, the volume decoder should generalize to the test distribution.

Reference: Ganin et al. 2016 "Domain-Adversarial Training of Neural Networks" (DANN), JMLR.

**Why this should work:** The stochastic subsampling experiment (PR #968) WIDENED the gap to +8.115pp (widest ever) — evidence the backbone is memorizing spatial point positions from the training distribution. DANN directly attacks this: by making backbone features domain-invariant, point-position memorization is penalized.

**Previous attempt:** PR #1034 was closed because the DANN implementation was never committed to git (0 diff vs base branch), preventing review and merge. This v2 PR is the clean replacement. W&B run `9q77j1wi` from PR #1034 may still be alive — you can continue monitoring it or launch a fresh run.

## Instructions to Student

**FIRST: Commit your DANN implementation to this branch immediately.** If you have working code on your pod filesystem, run:
```bash
git add target/model.py target/train.py target/trainer_runtime.py  # add any files you changed
git commit -m "feat: DANN domain adversarial training"
git push origin dl24-tanjiro/dann-domain-adversarial-v2
```
Then mark this PR ready for review.

### Code changes required

**1. Add `GradientReversalFunction` and `GradientReversalLayer` to `target/model.py`:**

```python
import torch
from torch.autograd import Function

class GradientReversalFunction(Function):
    @staticmethod
    def forward(ctx, x, alpha):
        ctx.save_for_backward(torch.tensor(alpha))
        return x.clone()

    @staticmethod
    def backward(ctx, grad_output):
        alpha, = ctx.saved_tensors
        return -alpha * grad_output, None

class GradientReversalLayer(nn.Module):
    def __init__(self, alpha=1.0):
        super().__init__()
        self.alpha = alpha

    def forward(self, x):
        return GradientReversalFunction.apply(x, self.alpha)
```

**2. Add a `DomainDiscriminator` head to `target/model.py`:**

```python
class DomainDiscriminator(nn.Module):
    def __init__(self, hidden_dim=512):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(hidden_dim, 256),
            nn.ReLU(),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, 1),  # binary: train=0, test=1
        )

    def forward(self, x):
        return self.net(x)
```

**3. Integrate into the main model class:** After the backbone encoder produces per-sample feature vectors (mean-pool along the point dimension to get a `[B, hidden_dim]` tensor), pass through GRL then DomainDiscriminator to get domain logits. The GRL alpha should be settable from outside (add `set_grl_alpha(alpha)` method).

**4. Add CLI flags to `target/train.py`:**
```python
parser.add_argument("--use-domain-adversarial", action="store_true", default=False,
                    help="Enable domain adversarial training (DANN)")
parser.add_argument("--domain-adv-lambda", type=float, default=0.1,
                    help="GRL reversal strength lambda")
parser.add_argument("--domain-adv-schedule", type=str, default="constant",
                    choices=["constant", "linear"],
                    help="Lambda schedule: constant or linearly ramp from 0 to domain-adv-lambda")
```

**5. Modify DataLoader to emit domain labels:**
- Training split samples → `domain_label = 0`
- Test split samples → `domain_label = 1`
- **Important:** Test split samples should be included in the training loop **without pressure targets** — they only contribute domain labels, not supervised loss. Create a separate unlabeled dataloader for test cases and interleave batches (or use a combined dataloader with a mask).

**6. Training step modification:**
```python
if args.use_domain_adversarial:
    # lambda schedule (optional ramp-up for stability)
    if args.domain_adv_schedule == "linear":
        p = global_step / total_steps
        lam = args.domain_adv_lambda * (2.0 / (1.0 + math.exp(-10 * p)) - 1)
    else:
        lam = args.domain_adv_lambda
    model.set_grl_alpha(lam)

    # Domain adversarial loss (all samples, train + test)
    domain_logits = model.get_domain_logits(backbone_features)  # [B]
    domain_labels = batch["domain_label"].float().to(device)    # [B]
    domain_loss = F.binary_cross_entropy_with_logits(domain_logits, domain_labels)

    # Total loss = supervised loss + domain loss (no GradNorm interaction needed)
    total_loss = supervised_loss + domain_loss
```

### Run configuration

Use the **exact wave SOTA config** as baseline, adding the domain adversarial flags:

```bash
cd target/ && python train.py \
  --model-layers 6 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --model-pe string_multisigma --pe-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --use-gradnorm --gradnorm-alpha 0.5 \
  --use-y-symmetry-aug --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65000 \
  --weight-decay 0.005 --lr 1e-4 \
  --epochs 30 \
  --no-compile-model \
  --lr-warmup-epochs 1 \
  --use-domain-adversarial \
  --domain-adv-lambda 0.1 \
  --domain-adv-schedule linear \
  --wandb_group val-test-gap-domain-adaptation
```

### Kill gates (val_primary/abupt_axis_mean_rel_l2_pct, operator `<` NOT `>`)
- EP5:  val_abupt < 7.5% to continue
- EP10: val_abupt < 7.2% to continue
- EP15: val_abupt < 6.80% to continue
- EP20: val_abupt < 6.70% to continue
- EP25: val_abupt < 6.65% to continue
- EP30: final evaluation — run test eval and post SENPAI-RESULT

### Implementation notes
- CRITICAL: `--no-compile-model` is REQUIRED (NCCL ALLREDUCE deadlock without it)
- CRITICAL: `--train-volume-points 65000` is REQUIRED
- CRITICAL: Use `--lr-warmup-epochs 1` NOT `--lr-warmup-steps 500`
- The domain discriminator does NOT interact with GradNorm — GradNorm handles the supervised task losses only; domain loss is added separately
- If lambda=0.1 is too aggressive (val_abupt degrades badly at EP5), try lambda=0.05 in a second arm
- W&B group: `val-test-gap-domain-adaptation`
- W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

### Expected outcome
Primary target: `test_primary/volume_pressure_rel_l2_pct` below **10.758%** (current wave SOTA).
Stretch target: approaching public SOTA of 6.08%.
The val→test vol_p gap should **narrow** — if it widens, that is a strong signal this approach is not working.

### Terminal result format
When the run completes or is killed, report:
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test_primary/volume_pressure_rel_l2_pct","value":<number>}}
```

## Baseline (wave SOTA)

PR #740, W&B run `5x8wofzm`

| Metric | Wave SOTA | Public SOTA Target |
|--------|-----------|-------------------|
| test_primary/abupt_axis_mean_rel_l2_pct | 7.5195% | — |
| test_primary/surface_pressure_rel_l2_pct | 3.8810% | 3.82% |
| **test_primary/volume_pressure_rel_l2_pct** | **10.7580%** | **6.08%** ← primary target |
| test_primary/wall_shear_rel_l2_pct | 7.0610% | 7.29% |

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

**Note:** If W&B run `9q77j1wi` from the previous PR is still active, include that run ID in your SENPAI-RESULT. If it has been killed or timed out, launch a fresh run with the command above.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Your DANN experiment has been re-opened as PR #1037 (clean branch, v2). PR #1034 was closed because no code was ever committed to git.

**Action required immediately:**
1. Check if run `9q77j1wi` is still alive on your pod
2. If still alive, commit your DANN implementation to this new branch:
   ```bash
   git fetch origin
   git checkout dl24-tanjiro/dann-domain-adversarial-v2
   git add target/model.py target/train.py  # add all files you changed for DANN
   git commit -m 'feat: DANN domain adversarial training'
   git push origin dl24-tanjiro/dann-domain-adversarial-v2
   ```
3. If run `9q77j1wi` has been killed (pod restart after PR closure), re-launch with the exact command in the PR body
4. Mark this PR ready for review once code is committed and run is live

**EP5 kill gate:** val_abupt ≤ 7.5% at step ~54,875

If your pod restarted and the DANN code was lost, re-implement it from scratch using the detailed instructions in the PR body — it's straightforward and you did it correctly the first time (domain_loss ≈ 0.693 confirmed working in run `9q77j1wi`). — Advisor

**Comment by morganmcg1** [comment]
ADVISOR: URGENT — Run `9q77j1wi` is confirmed alive (step=17,213, EP~1.6, domain_loss working), but your DANN code is STILL not committed to this branch. You have ~2 hours before the EP5 kill gate (step ~54,930, val_abupt ≤ 7.5%). 

If you don't commit the code before the gate, the run cannot be merged regardless of the metric result — we need reproducible code in git.

**Do this right now:**
```bash
git fetch origin
git checkout dl24-tanjiro/dann-domain-adversarial-v2
git add target/model.py target/train.py
git commit -m 'feat: DANN domain adversarial training with GRL'
git push origin dl24-tanjiro/dann-domain-adversarial-v2
```

The run is still running so your pod survived. Commit the code immediately. — Advisor

**Comment by morganmcg1** [comment]
ADVISOR: W&B update on run `9q77j1wi` (direct check, EP~1.6):

| Metric | Value |
|--------|-------|
| Step | 17,251 |
| val_abupt | 11.8265% |
| val_surf_p | 8.2434% |
| val_vol_p | 9.6448% |
| val_wss | 12.3866% |
| Slope (abupt) | **-0.9954%/1k_steps** |

The val metrics are high at EP1 but the slope is excellent (-0.9954%/1k_steps). If this slope holds, by EP5 (~step 54,930) you'd expect val_abupt to drop by ~37pp from here — that's clearly not linear, but the steep early learning is encouraging.

**HOWEVER**: your DANN code is still NOT committed to this PR branch. You have approximately 2 hours before the EP5 gate check at step ~54,930. Without code committed, this run cannot be merged.

**THIS IS YOUR LAST WARNING before I close this PR.** If the code is not committed to `dl24-tanjiro/dann-domain-adversarial-v2` within the next 30 minutes, I will close this PR for the same reason as #1034 — untracked code.

Please post EP2 metrics when the eval completes (step ~21,972). — Advisor

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1037 because ADVISOR: Closing PR #1037. Despite three escalation comments over the past hour (13:59Z, 14:19Z, 14:19Z) and an explicit 30-minute last warning, no DANN implementation code has been committed to this branch. PR #1034 was closed for the same reason, and PR #1037 was a clean-slate second chance that has also been forfeited.

Run `9q77j1wi` may still be running on the pod, but without reproducible code in git there is nothing to merge regardless of metric results.

If you still want to run this experiment: implement DANN from scratch using the instructions in the PR body, start a fresh branch, and open a new PR. The code commitment must happen BEFORE or SIMULTANEOUSLY with launching the run. Do not launch a run and commit code later. — Advisor.
